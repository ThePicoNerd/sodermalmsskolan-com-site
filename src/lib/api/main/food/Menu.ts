import useSWR from "swr";
import { fetchJSON } from "../fetch";
import { CollectionResponse } from "../Response";
import moment from "moment";

export interface Dish {
  title: string;
  id: string;
}

export interface Menu {
  dishes: Dish[];
  date: Date;
}

export type MenuResponse = CollectionResponse<Menu>;

export interface MenuQuery {
  /**
   * How many menus to retreive (maximum).
   */
  limit?: number;

  /**
   * Offset the starting day, counting from today.
   */
  offset?: number;
}

/**
 *
 */
export function useMenus({ limit = 10, offset = 0 }: MenuQuery) {
  return useSWR(`/api/food/menus?limit=${limit}&offset=${offset}`, async () => {
    const res = await fetchJSON<MenuResponse>("/api/food/menus");

    const menus = res.data.sort(({ date: a }, { date: b }) => {
      if (a < b) {
        return -1;
      }

      if (a > b) {
        return 1;
      }

      return 0;
    });

    const now = moment(new Date());
    const next = menus.findIndex((menu) =>
      now.isSameOrBefore(menu.date, "date")
    );
    const startIndex = next + offset;

    return menus.slice(startIndex, startIndex + limit);
  });
}
