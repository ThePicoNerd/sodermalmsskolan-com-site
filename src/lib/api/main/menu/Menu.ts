import { CollectionResponse } from "../Response";
import useSWR from "swr";
import { fetchJSON } from "../fetch";

export interface Menu {
  dishes: string[];
  timestamp: Date;
}

export type MenuResponse = CollectionResponse<Menu>;

export interface MenuQuery {
  /**
   * How many menus to retreive (maximum).
   */
  limit?: number;

  /**
   * Offset the starting day, counting from today.
   */
  offset?: number;
}

/**
 *
 */
export function useMenus({ limit = 10, offset = 0 }: MenuQuery) {
  return useSWR(
    `/api/menus?limit=${limit}&offset=${offset}`,
    async (url: string) => {
      const res = await fetchJSON<MenuResponse>(url);
      return res.data;
    }
  );
}
