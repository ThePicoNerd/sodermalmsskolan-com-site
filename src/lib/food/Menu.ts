export type Dishes = string[];

export default interface Menu {
  date: string;
  dishes: Dishes;
}
