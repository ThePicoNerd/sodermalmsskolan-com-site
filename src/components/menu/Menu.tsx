import dayjs from "dayjs";
import React, { FunctionComponent } from "react";
import { useLang } from "../../hooks/lang";
import { ClientMenu } from "../../lib/food/structures/client/Menu";
import DishCard from "../dishes/DishCard";
import CardGrid from "../grid/CardGrid";
import { SmallHeading } from "../text/headings";

const Menu: FunctionComponent<{menu: ClientMenu}> = ({ menu }) => {
  const lang = useLang();

  return (
    <div css={{
      display: "flex",
    }}
    >
      <SmallHeading
        css={{
          writingMode: "vertical-rl",
          marginRight: "1rem",
        }}
      >
        {menu?.date ? dayjs(menu?.date).locale(lang).format("dddd DD MMMM") : "LÃ¤ser in ..."}
      </SmallHeading>
      <CardGrid css={{
        flex: 1,
      }}
      >
        {(menu?.dishes || new Array(2).fill(null))
          .map((dish, index) => <DishCard dish={dish} key={dish?.id || index} />)}
      </CardGrid>
    </div>
  );
};

export default Menu;
