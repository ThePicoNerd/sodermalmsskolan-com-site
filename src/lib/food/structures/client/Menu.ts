import useSWR, { responseInterface } from "swr";
import { IMenu, Menu, MenuQuery } from "../shared/Menu";
import ky from "ky-universal";
import dayjs from "dayjs";
import { ClientDish } from "./Dish";
import { useLocale } from "../../../../hooks/locale";
import { firstLetterUpperCase } from "../../../utils/letters";
import isSameOrBefore from "dayjs/plugin/isSameOrBefore";

dayjs.extend(isSameOrBefore);

export class ClientMenu extends Menu {
  dishes: ClientDish[];

  constructor({ dishes, date }: IMenu) {
    super({ dishes, date });

    this.dishes = dishes.map((dish) => new ClientDish(dish));
  }

  public static async fetchAll(): Promise<ClientMenu[]> {
    const res = await ky.get("/api/food/menus").json<IMenu[]>();

    const menus = res
      .map((menu) => new ClientMenu(menu))
      .sort((a, b) => a.date.getTime() - b.date.getTime());

    return menus;
  }

  public static use({
    limit = 10,
    offset = 0,
  }: MenuQuery): responseInterface<ClientMenu[], unknown> {
    return useSWR(
      `/api/food/menus?limit=${limit}&offset=${offset}`,
      async () => {
        const menus = await ClientMenu.fetchAll();

        const now = dayjs(new Date());
        const next = menus.findIndex((menu) =>
          now.isSameOrBefore(menu.date, "date")
        );
        const startIndex = next + offset;

        return menus.slice(startIndex, startIndex + limit);
      }
    );
  }

  public get title(): string {
    const { locale } = useLocale();

    return firstLetterUpperCase(
      dayjs(this?.date).locale(locale).format("dddd D MMMM")
    );
  }
}
