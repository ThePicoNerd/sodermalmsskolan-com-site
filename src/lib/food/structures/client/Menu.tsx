import useSWR from "swr";
import ky from "ky-universal";
import dayjs from "dayjs";
import isSameOrBefore from "dayjs/plugin/isSameOrBefore";
import { ClientDish } from "./Dish";
import { IMenu, Menu } from "../shared/Menu";
import { UseSWRResource } from "../../../common/usable";

dayjs.extend(isSameOrBefore);

export interface UseMenuQuery {
  limit?: number;
  offset?: number;
}

export class ClientMenu extends Menu {
  dishes: ClientDish[];

  constructor({ dishes, date }: IMenu) {
    super({ dishes, date });

    this.dishes = dishes.map((dish) => new ClientDish(dish));
  }

  public static async fetchAll(): Promise<ClientMenu[]> {
    const res = await ky.get("/api/food/menus").json<IMenu[]>();

    const menus = res
      .map((menu) => new ClientMenu(menu))
      .sort((a, b) => a.date.getTime() - b.date.getTime());

    return menus;
  }
}

export const useMenu: UseSWRResource<ClientMenu[], UseMenuQuery> = ({
  limit = 10,
  offset = 0,
}) => useSWR(`/api/food/menus?limit=${limit}&offset=${offset}`, async () => {
  const menus = await ClientMenu.fetchAll();

  const now = dayjs(new Date());
  const next = menus.findIndex((menu) => now.isSameOrBefore(menu.date, "date"));
  const startIndex = next + offset;

  return menus.slice(startIndex, startIndex + limit);
});
