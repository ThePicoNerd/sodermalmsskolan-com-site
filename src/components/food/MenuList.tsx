import styled from "styled-components";
import React from "react";
import moment from "moment";
import Skeleton from "react-loading-skeleton";
import { Row } from "../grid/Row";
import { Col } from "../grid/Col";
import { firstLetterUpperCase } from "../../lib/utils/letters";
import * as breakpoints from "../../styles/breakpoints";
import { Menu } from "../../lib/food/structures/shared/Menu";
import { ClientMenu } from "../../lib/food/structures/client/Menu";
import { Dish } from "../../lib/food/structures/shared/Dish";
import { ClientDish } from "../../lib/food/structures/client/Dish";
import { useLocale } from "../../hooks/locale";

const ItemTitle = styled.h6`
  margin-bottom: 1rem;
`;

const DishEmissions: React.FunctionComponent<{ id: string }> = ({ id }) => {
  const { data } = ClientDish.use(id);
  const { locale } = useLocale();

  return (
    <span>
      ({data?.co2e.toLocaleString(locale) || <Skeleton width={32} />} kg CO₂e
      per kg)
    </span>
  );
};

const DishItem: React.FunctionComponent<{ dish: Dish }> = ({ dish }) => {
  return (
    <li>
      {dish.title} <DishEmissions id={dish?.id} />
    </li>
  );
};

const MenuListItem: React.FunctionComponent<{ menu: Menu }> = ({ menu }) => {
  const { locale } = useLocale();
  const date = menu?.date ? (
    firstLetterUpperCase(
      moment(menu?.date).locale(locale).format("dddd D MMMM")
    )
  ) : (
    <Skeleton />
  );

  return (
    <div>
      <ItemTitle>{date}</ItemTitle>
      <ul>
        {menu?.dishes?.map((dish, index) => (
          <DishItem key={index} dish={dish} />
        ))}
      </ul>
    </div>
  );
};

const Grid = styled.div`
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: var(--grid-gap);

  @media (min-width: ${breakpoints.small}) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media (min-width: ${breakpoints.large}) {
    grid-template-columns: repeat(3, 1fr);
  }
`;

export const MenuList: React.FunctionComponent<{
  limit: number;
}> = ({ limit }) => {
  const { data, isValidating } = ClientMenu.use({ limit });
  const fallbackArray: Menu[] = new Array(limit).fill(null);
  const menus = data?.length > 0 ? data : fallbackArray;
  const isEmpty = !isValidating && !data;

  return (
    <Row>
      <Col xs={12} md={9} lg={8}>
        <Grid>
          {isEmpty ? (
            <p>Menyn är inte tillgänglig.</p>
          ) : (
            menus.map((menu, index) => <MenuListItem key={index} menu={menu} />)
          )}
        </Grid>
      </Col>
    </Row>
  );
};
