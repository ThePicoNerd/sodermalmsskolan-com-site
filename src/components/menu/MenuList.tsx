import React, { FunctionComponent } from "react";
import { AlertOctagon } from "react-feather";
import useMenus from "../../lib/food/hooks/useMenus";
import Menu from "../../lib/food/Menu";
import { useMenuContext } from "../../lib/food/MenuContext";
import MenuDisplay from "./MenuDisplay";

export interface MenuListDisplayProps {
  menus: Menu[];
  expectedCount: number;
}

/**
 * Display the glorious menus.
 *
 * @param {React.PropsWithChildren<MenuListDisplayProps>} props Props.
 *
 * @returns {React.ReactElement} The rendered list.
 */
export const MenuListDisplay: FunctionComponent<MenuListDisplayProps> = ({
  menus,
  expectedCount,
}) => {
  if (menus?.length === 0) {
    return (
      <div className="my-4 text-gray-400 dark:text-gray-500">
        <AlertOctagon className="mx-auto w-16 h-16 text-gray-300 dark:text-gray-700 mb-2" strokeWidth="1" />
        <p className="text-center leading-tight">Ingen meny hittades.</p>
      </div>
    );
  }

  return (
    <ul className="my-4">
      {(menus ?? new Array(expectedCount).fill(undefined)).map((menu, i) => (
        <li key={menu?.date ?? i}>
          <MenuDisplay menu={menu} />
        </li>
      ))}
    </ul>
  );
};

/**
 * A list of menus.
 *
 * @returns {React.ReactElement} The rendered list.
 */
const MenuList: FunctionComponent = () => {
  const { cursor, scope } = useMenuContext();

  const first = cursor?.startOf(scope);
  const last = cursor?.endOf(scope);

  const menus = useMenus({ first, last });

  const expectedCount = Math.ceil((last?.diff(first, "days")?.days ?? 0) * (5 / 7)); // Usually, school is closed on weekends.

  return (
    <MenuListDisplay menus={menus} expectedCount={expectedCount} />
  );
};

export default MenuList;
