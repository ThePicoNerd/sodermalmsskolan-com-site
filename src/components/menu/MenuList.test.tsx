import { render, screen } from "@testing-library/react";
import React from "react";
import Menu from "../../lib/food/Menu";
import { MenuListDisplay } from "./MenuList";

describe("<MenuList /> tests", () => {
  it("should return some helpful text if there are no menus", () => {
    const { asFragment } = render(<MenuListDisplay menus={[]} expectedCount={5} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display all of the menus", () => {
    const menus: Menu[] = [{
      date: "2020-01-01",
      dishes: ["Tacobuffé", "Fisk Björkeby"],
    }, {
      date: "2020-01-02",
      dishes: ["Pannkakor", "Pizza"],
    }, {
      date: "2020-01-03",
      dishes: ["Hamburgare", "Köttbullar"],
    }];

    render(<MenuListDisplay menus={menus} expectedCount={0} />);

    menus.flatMap((menu) => menu.dishes).forEach((dish) => {
      expect(screen.getByText(dish)).toBeDefined();
    });
  });

  it("should fall back to skeletons", () => {
    const { asFragment } = render(<MenuListDisplay menus={undefined} expectedCount={5} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
