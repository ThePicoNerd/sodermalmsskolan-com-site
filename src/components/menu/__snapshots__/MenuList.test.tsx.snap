// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuList /> tests should fall back to skeletons 1`] = `
<DocumentFragment>
  <ul
    class="my-4"
  >
    <li>
      <div
        class="my-4"
      >
        <h3
          class="text-xl font-semibold"
        >
          <span>
            <span
              class="skeleton max-w-full"
              style="width: 12em;"
            >
              ‌
            </span>
          </span>
        </h3>
        <ul>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li>
      <div
        class="my-4"
      >
        <h3
          class="text-xl font-semibold"
        >
          <span>
            <span
              class="skeleton max-w-full"
              style="width: 12em;"
            >
              ‌
            </span>
          </span>
        </h3>
        <ul>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li>
      <div
        class="my-4"
      >
        <h3
          class="text-xl font-semibold"
        >
          <span>
            <span
              class="skeleton max-w-full"
              style="width: 12em;"
            >
              ‌
            </span>
          </span>
        </h3>
        <ul>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li>
      <div
        class="my-4"
      >
        <h3
          class="text-xl font-semibold"
        >
          <span>
            <span
              class="skeleton max-w-full"
              style="width: 12em;"
            >
              ‌
            </span>
          </span>
        </h3>
        <ul>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li>
      <div
        class="my-4"
      >
        <h3
          class="text-xl font-semibold"
        >
          <span>
            <span
              class="skeleton max-w-full"
              style="width: 12em;"
            >
              ‌
            </span>
          </span>
        </h3>
        <ul>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
          <li
            class="text-gray-500 font-medium leading-snug my-1"
          >
            <span>
              <span
                class="skeleton max-w-full"
                style="width: 40em;"
              >
                ‌
              </span>
            </span>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<MenuList /> tests should return some helpful text if there are no menus 1`] = `
<DocumentFragment>
  <div
    class="my-4 text-gray-400 dark:text-gray-500"
  >
    <svg
      class="mx-auto w-16 h-16 text-gray-300 dark:text-gray-700 mb-2"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.86 2 16.14 2 22 7.86 22 16.14 16.14 22 7.86 22 2 16.14 2 7.86 7.86 2"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="12"
      />
      <line
        x1="12"
        x2="12.01"
        y1="16"
        y2="16"
      />
    </svg>
    <p
      class="text-center leading-tight"
    >
      Ingen meny hittades.
    </p>
  </div>
</DocumentFragment>
`;
